src/Psl/Type/README.md

```diff
@@ -0,0 +1,1281 @@
+): TypeInterface<iterable<Tk, Tv>>
+```
+
+Provides a type that can parse iterables. The `Tv` type can consist out of advanced types like shapes, vectors, or other dictionaries.
+
+Can coerce from:
+
+* `iterable`: The iterable will be iterated:
+    * Every item's key will be coerced by the provided `Tk` type.
+    * Every item's value will be coerced by the provided `Tv` type.
+
+These are some examples on how the type can be used:
+
+```php
+use Psl\Type;
+
+$iterable = Type\iterable(
+    Type\string(),
+    Type\shape([
+        'title' => Type\string(),
+        'content' => Type\string(),
+    ])
+);
+
+$iterable->assert([
+    'string-key' => [
+        'title' => 'Hello world',
+        'content' => 'A book that greets the world in all possible languages!',
+    ]
+]);
+```
+
+When the iterable does not match the specified type, you will get detailed information about what went wrong exactly:
+
+> Expected "iterable<string, array{'title': string, 'content': string}>", got "int" **at path "key(123)"**.
+
+> Expected "iterable<string, array{'title': string, 'content': string}>", got "int" **at path "foo.title"**.
+
+---
+
+#### [literal_scalar](literal_scalar.php)
+
+```hack
+@pure
+@template T of string|int|float|bool
+Type\literal_scalar(T): TypeInterface<T>
+```
+
+Provides a type that can parse a literal scalar value.
+
+Can coerce from:
+
+* `T`: The provided value will be compared against the expected value. If it matches, the coercion has succeeded.
+* `string`: The provided value will be coerced into a string and compared against the expected value.
+* `int`: The provided value will be coerced into an integer and compared against the expected value.
+* `float`: The provided value will be coerced into a float and compared against the expected value.
+* `bool`: The provided value will be coerced into a boolean and compared against the expected value.
+
+Examples:
+
+```php
+use Psl\Type;
+
+Type\literal_scalar('hello')->assert('hello');
+```
+
+#### [map](map.php)
+
+```hack
+@pure
+@template Tk of array-key
+@template Tv of mixed
+Type\map(
+    TypeInterface<Tk> $key_type,
+    TypeInterface<Tv> $value_type
+): TypeInterface<MapInterface<Tk, Tv>>
+```
+
+Provides a type that can parse into a `Psl\Collection\MapInterface`. The `Tv` type can consist out of advanced types like shapes, vectors, or other dictionaries.
+
+Can coerce from:
+
+* `iterable`: The iterable will be iterated:
+    * Every item's key will be coerced by the provided `Tk` type.
+    * Every item's value will be coerced by the provided `Tv` type.
+
+These are some examples on how the type can be used:
+
+```php
+use Psl\Type;
+
+$map = Type\map(
+    Type\string(),
+    Type\shape([
+        'title' => Type\string(),
+        'content' => Type\string(),
+    ])
+);
+
+$map->assert([
+    'string-key' => [
+        'title' => 'Hello world',
+        'content' => 'A book that greets the world in all possible languages!',
+    ]
+]);
```

Comment:
This example will give an error, it should be either `coerce` or provide an instance of `MapInterface`.